src/Device/Driver/AirControlDisplay.cpp

```diff
@@ -123,6 +127,226 @@ ACDDevice::ParseNMEA(const char *_line, NMEAInfo &info)
+ * 13) Age of differential GPS data, time in seconds since last SC104
+ *     type 1 or 9 update, null field when DGPS is not used
+ * 14) Differential reference station ID, 0000-1023
+ * 15) Checksum
+ */
+static bool
+FormatGPGGA(char *buffer, size_t buffer_size, const MoreData &info)
+{
+  char lat_buffer[20];
+  char long_buffer[20];
+
+  const GeoPoint location = info.location_available
+    ? info.location
+    : GeoPoint::Zero();
+
+  FormatLatitude(lat_buffer, sizeof(lat_buffer), location.latitude);
+  FormatLongitude(long_buffer, sizeof(long_buffer), location.longitude);
+
+  const BrokenDateTime now = info.time_available &&
+    info.date_time_utc.IsDatePlausible()
+    ? info.date_time_utc
+    : BrokenDateTime::NowUTC();
+
+  snprintf(buffer, buffer_size,
+           "GPGGA,%02u%02u%02u,%c,%s,%s,%u,%2u,%.1f,%.1f,%c,%.1f,%c,,",
+           now.hour, now.minute, now.second,
+           info.location.IsValid() ? 'A' : 'V',
+           lat_buffer,
+           long_buffer,
+           (int)info.gps.fix_quality,
+           info.gps.satellites_used,
+           info.gps.hdop,
+           info.gps_altitude,
+           'M',
+           info.gps_altitude,
+           'M');
+
+  return true;
+}
+
+/*
+ * $GPGSA,a,a,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x.x,x.x,x.x*hh
+ *
+ * Field Number:
+ *  1) Selection mode
+ *         M=Manual, forced to operate in 2D or 3D
+ *         A=Automatic, 3D/2D
+ *  2) Mode (1 = no fix, 2 = 2D fix, 3 = 3D fix)
+ *  3) ID of 1st satellite used for fix
+ *  4) ID of 2nd satellite used for fix
+ *  ...
+ *  14) ID of 12th satellite used for fix
+ *  15) PDOP
+ *  16) HDOP
+ *  17) VDOP
+ *  18) checksum
+ */
+static bool
+FormatGPGSA(char *buffer, size_t buffer_size, const MoreData &info)
+{
+  // don't format GPGSA string if no sat id's are available
+  if(!info.gps.satellite_ids_available)
+    return false;
+
+  int gps_status;
+
+  if (!info.alive || !info.location_available)
+    gps_status = 1;
+  else if (!info.gps_altitude_available)
+    gps_status = 2;
+  else
+    gps_status = 3;
+
+  snprintf(buffer, buffer_size,
+           "GPGSA,A,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%.1f,%.1f,%.1f",
+           gps_status,
+           info.gps.satellite_ids[0],
+           info.gps.satellite_ids[1],
+           info.gps.satellite_ids[2],
+           info.gps.satellite_ids[3],
+           info.gps.satellite_ids[4],
+           info.gps.satellite_ids[5],
+           info.gps.satellite_ids[6],
+           info.gps.satellite_ids[7],
+           info.gps.satellite_ids[8],
+           info.gps.satellite_ids[9],
+           info.gps.satellite_ids[10],
+           info.gps.satellite_ids[11],
+           info.gps.pdop,
+           info.gps.hdop,
+           info.gps.vdop);
+
+  return true;
+}
+
+void
+ACDDevice::OnCalculatedUpdate(const MoreData &basic,
+                        const DerivedInfo &calculated)
+{
+  NullOperationEnvironment env;
+  char buffer[100];
+
+  if (FormatGPRMC(buffer, sizeof(buffer), basic))
```

Comment:
Only send this if a new GPS location has been received; imagine one device has a 20 Hz pressure sensor, but doesn't provide location data; this would trigger sending those sentences 20 times a second.

Reply:
I agree, makes sense.

Do you think I will still have a freezing app when would have a 20 Hz GPS receiver and use `OnSensorUpdate()`?